pkg/server/plugin/upstreamauthority/gcpca/gcpca.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+func (p *Plugin) mintX509CA_(log hclog.Logger, ctx context.Context, csr []byte, preferredTtl int32) (*upstreamauthority.MintX509CAResponse, error) {
+	csrParsed, err := x509.ParseCertificateRequest(csr)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse CSR: %v", err)
+	}
+
+	validity := time.Second * time.Duration(preferredTtl)
+	parentPath := "projects/" + p.c.Project + "/locations/" + p.c.Location + "/certificateAuthorities/" + p.c.CertName
+
+	log.Info("Minting X509 using certificate at " + parentPath + "; ttl " + validity.String())
+
+	subject := privatecapb.Subject{}
+	extractFirst := func(strings []string, into *string) {
+		if len(strings) > 0 {
+			*into = strings[0]
+		}
+	}
+
+	extractFirst(csrParsed.Subject.Organization, &subject.Organization)
+	extractFirst(csrParsed.Subject.OrganizationalUnit, &subject.OrganizationalUnit)
+	extractFirst(csrParsed.Subject.Locality, &subject.Locality)
+	extractFirst(csrParsed.Subject.Province, &subject.Province)
+
+	createRequest := privatecapb.CreateCertificateRequest{
+		Parent: parentPath,
+		Certificate: &privatecapb.Certificate{
+			Lifetime: ptypes.DurationProto(validity),
+			CertificateConfig: &privatecapb.Certificate_Config{
+				Config: &privatecapb.CertificateConfig{
+					PublicKey: &privatecapb.PublicKey{
+						Type: privatecapb.PublicKey_PEM_EC_KEY,
+						Key: pem.EncodeToMemory(
+							&pem.Block{
+								Type:  "PUBLIC KEY",
+								Bytes: csrParsed.RawSubjectPublicKeyInfo,
+							},
+						),
+					},
+					SubjectConfig: &privatecapb.CertificateConfig_SubjectConfig{
+						Subject:    &subject,
+						CommonName: csrParsed.Subject.CommonName,
+					},
+					ReusableConfig: &privatecapb.ReusableConfigWrapper{
+						ConfigValues: &privatecapb.ReusableConfigWrapper_ReusableConfig{
+							ReusableConfig: fmt.Sprintf("projects/%s/locations/%s/reusableConfigs/%s", "privateca-data", p.c.Location, "subordinate-mtls-pathlen-0"),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	pcaClient, err := getClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cresp, err := pcaClient.CreateCertificate(ctx, &createRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := pemToX509(cresp.GetPemCertificate())
+	if err != nil {
+		return nil, err
+	}
+
+	certChain := make([]*x509.Certificate, len(cresp.PemCertificateChain))
+	for i, c := range cresp.PemCertificateChain {
+		certChain[i], err = pemToX509(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// The last certificate returned from the chain is the root.
+	upstreamRoot := certChain[len(certChain)-1]
+	bundle := x509util.DedupeCertificates([]*x509.Certificate{upstreamRoot})
+
+	derBundle := x509util.RawCertsFromCertificates(bundle)
+
+	// All else comprises the chain (including the issued certificate)
+	fullChain := []*x509.Certificate{cert}
+	fullChain = append(fullChain, certChain...)
```

Comment:
The `X509CaChain` should only contain the newly minted intermediate CA and then other intermediates necessary to chain it back to a root present in the `UpstreamX509Roots` field. My reading of this code is that the upstream root is present in `certChain` and would therefore be appended to `X509CaChain`, which is not correct. `X509CaChain` should not contain the root.

Reply:
Fixed thank you